src/libsystemd/sd-login/sd-login.c

```diff
@@ -134,6 +135,206 @@ _public_ int sd_pid_get_cgroup(pid_t pid, char **cgroup) {
         return 0;
 }
 
+_public_ int sd_pidfd_get_session(int pidfd, char **session) {
+        pid_t pid_early, pid_late;
+        int r;
+
+        assert_return(pidfd >= 0, -EBADF);
+        assert_return(session, -EINVAL);
+
+        r = pidfd_get_pid(pidfd, &pid_early);
+        if (r < 0)
+                return r;
+
+        r = cg_pid_get_session(pid_early, session);
+        if (r < 0)
+                return IN_SET(r, -ENXIO, -ENOMEDIUM) ? -ENODATA : r;
+
+        r = pidfd_get_pid(pidfd, &pid_late);
+        if (r < 0)
+                return r;
+
+        if (pid_early != pid_late)
+                return -ESRCH;
```

Comment:
i think it would make sense to add a helper like this:

```c
int pidfd_verify_pid(int pidfd, pid_t pid) {
        pid_t current_pid;
        int r;

        r = pidfd_get_pid(pidfd, &current_pid);
        if (r < 0)
                return r;
        if (current_pid != pid)
                return -ESRCH;
        return 0;
}
```
that would shorten the code a few lines, and ensure we always return the error as ESRCH.